### ISSUE TITLE
Printf floating point tests do not consider device rounding mode

### ISSUE THREAD

**Issue by Jeremy-Kemp** [KhronosGroup/OpenCL-CTS#673]
The reference results for the `printf` test are hard-coded. 

For implementations not implementing `CL_FP_ROUND_TO_NEAREST`, floating point tests can appear to fail.

For example, the first float test `printf("%f",10.3456)` expects a result of `10.345600`.  https://github.com/KhronosGroup/OpenCL-CTS/blob/8a00f4ecede75f15a822cfdd0321ad69e3acea8f/test_conformance/printf/util_printf.cpp#L238

On an implementation with `CL_FP_ROUND_TO_ZERO` the correct result is actually `10.345599`.

**Comment by Jeremy-Kemp** [comment]
@kpet can we start utilising Labels so that we can filter out serious issues/PR's from the not-so-serious? 

My first though to solve this is by using `std::stringstream` to generate the reference result (with the correct rounding mode set that matches the device). 

I'd appreciate some input about how we want to solve this.

Below snippet for clarity:

```c
#include <cfenv>
#include <string>  
#include <iostream> 
#include <sstream>   
#include <iomanip>

int main()
{
	// Save the current rounding mode
	const int old_round = std::fegetround();

	// Match this to whatever CL_DEVICE_SINGLE_FP_CONFIG, CL_DEVICE_SINGLE_FP_CONFIG or CL_DEVICE_HALF_FP_CONFIG is
	std::fesetround(FE_TOWARDZERO);

	// Build the reference result
	std::stringstream test1;
	test1 << std::fixed << std::setprecision(6) << 10.3456 << std::endl;

	// Restore rounding mode
	std::fesetround(old_round);

	// Save the reference result
	std::string reference = test1.str();

	// Print the reference result
	// Will print 10.345599 in this example
	std::cout <<  reference;

	return 0;
}
```

**Comment by kpet** [comment]
@Jeremy-Kemp Yes, I'm all for using the labels. I've created one for cleanups. I'll start labeling some issues. This one is a bug.

On this particular issue, what you're proposing seems reasonable at first glance. Tagging @jlewis-austin and @EwanC who've worked on several FP-related issues.

**Comment by jlewis-austin** [comment]
It seems reasonable, but with a few strings to pull wrt what the specs say on the subject. One concern is that some platforms could ignore the currently set rounding mode in order to provide exact round-trip behavior for printf--we might consider accepting the correctly-rounded answer in any case. 
We'd want to use the utilities in harness/rounding_mode.c for setting/getting rounding modes.

**Comment by Jeremy-Kemp** [comment]
> One concern is that some platforms could ignore the currently set rounding mode

Is that something we'd need to consider though given that that version of `printf` would be deviating from C99? Regardless, it'd leave us in no worse a position that we currently are in.

> We'd want to use the utilities in harness/rounding_mode.c for setting/getting rounding modes.

Yes, definitely. The code snippet was just a quick proof of concept.

@kpet thanks for the label (and doing assigning them to lots of other bugs too)

**Comment by EwanC** [comment]
No objections to the proof of concept snippet as a way to address this bug. Happy to help review a MR following on from it.

**Comment by jlewis-austin** [comment]
> Is that something we'd need to consider though given that that version of `printf` would be deviating from C99? Regardless, it'd leave us in no worse a position than we currently are in.
> 
That was one of the strings I wanted to pull--if that's what it says then that's fine. I was thinking of the case where the device and host modes differ. If the conversions are done on the host, I was assuming the round to nearest host implementation would be sufficient regardless of the device setting. It's been a couple years since I looked at printf stuff so apologies if I'm missing the boat here.

**Comment by Jeremy-Kemp** [comment]
I don't have the exact wording from C99 to check this. Is anyone else able to check the exact wording of printf rounding behaviour?

The fix is a bit more involved than I'd originally thought. Solving it generically will be difficult due to the test calling OpenCL builtins. For example, we can't generate the reference result for `printf ("%f\n", acospi(2.0f))` on the CPU generically using the existing test structure, it'd require special casing.

**Comment by jlewis-austin** [comment]
I don't have the official specs, but from what I've found in C99 drafts and IEEE 754-1985, it looks like printf allows using a non-RNE mode, but doesn't require it. So hypothetically a round-to-zero device could use the host printf without modifying the host rounding mode. I don't know if that situation exists in any implementations, but I agree we can cross that bridge when and if we come to it. 
Note that this is based on parsing a few key phrases, some of which appear to be worded differently in C11 and 754-2008.

**Comment by Jeremy-Kemp** [comment]
Closed via #675
